### ISSUE TITLE
Email polling. - 'Unable to load BODYSTRUCTURE.' 

### ISSUE THREAD

**Issue by ronankeane** [moqui/moqui-framework#97]
javax.mail.MessagingException: Unable to load BODYSTRUCTURE

Preceeded by a java.io.UnsupportedEncodingException: unicode-1-1-utf-7 error (that was resolved by adding jcharset-2.0.jar to jre/lib/ext). 

Full log for javax.mail.MessagingException error:

```
12:05:05.796  WARN oquiWorker-2      o.moqui.i.c.TransactionFacadeImpl Transaction rollback. The rollback was originally caused by: Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)
javax.mail.MessagingException: Unable to load BODYSTRUCTURE
        at com.sun.mail.imap.IMAPMessage.loadBODYSTRUCTURE(IMAPMessage.java:1488) ~[javax.mail-1.5.6.jar:1.5.6]
        at com.sun.mail.imap.IMAPMessage.getDataHandler(IMAPMessage.java:746) ~[javax.mail-1.5.6.jar:1.5.6]
        at javax.mail.internet.MimeMessage.getContent(MimeMessage.java:1454) ~[javax.mail-1.5.6.jar:1.5.6]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:124) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.runIfMatches(EmailEcaRule.groovy:84) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceFacadeImpl.runEmecaRules(ServiceFacadeImpl.groovy:429) ~[moqui-framework-2.0.0.jar:2.0.0]
        at pollEmailServer_groovy.run(pollEmailServer_groovy:89) ~[script:?]
        at org.moqui.impl.context.runner.GroovyScriptRunner.run(GroovyScriptRunner.groovy:50) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.ResourceFacadeImpl.script(ResourceFacadeImpl.groovy:356) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.runner.ScriptServiceRunner.runService(ScriptServiceRunner.groovy:49) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:317) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
12:05:05.814  WARN oquiWorker-2      o.moqui.i.c.TransactionFacadeImpl Transaction rollback for [Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)]. Here is the current location: 
org.moqui.BaseException: Rollback location
        at org.moqui.impl.context.TransactionFacadeImpl.rollback(TransactionFacadeImpl.groovy:425) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.TransactionFacadeImpl.rollback(TransactionFacadeImpl.groovy:408) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:340) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
12:05:05.915  WARN oquiWorker-2        o.moqui.i.s.ServiceCallSyncImpl Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)
javax.mail.MessagingException: Unable to load BODYSTRUCTURE
        at com.sun.mail.imap.IMAPMessage.loadBODYSTRUCTURE(IMAPMessage.java:1488) ~[javax.mail-1.5.6.jar:1.5.6]
        at com.sun.mail.imap.IMAPMessage.getDataHandler(IMAPMessage.java:746) ~[javax.mail-1.5.6.jar:1.5.6]
        at javax.mail.internet.MimeMessage.getContent(MimeMessage.java:1454) ~[javax.mail-1.5.6.jar:1.5.6]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:124) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.runIfMatches(EmailEcaRule.groovy:84) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceFacadeImpl.runEmecaRules(ServiceFacadeImpl.groovy:429) ~[moqui-framework-2.0.0.jar:2.0.0]
        at pollEmailServer_groovy.run(pollEmailServer_groovy:89) ~[script:?]
        at org.moqui.impl.context.runner.GroovyScriptRunner.run(GroovyScriptRunner.groovy:50) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.ResourceFacadeImpl.script(ResourceFacadeImpl.groovy:356) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.runner.ScriptServiceRunner.runService(ScriptServiceRunner.groovy:49) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:317) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
```

Full log for java.io.UnsupportedEncodingException error:

```
11:33:57.544  WARN oquiWorker-3      o.moqui.i.c.TransactionFacadeImpl Transaction rollback. The rollback was originally caused by: Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)
java.io.UnsupportedEncodingException: unicode-1-1-utf-7
        at sun.nio.cs.StreamDecoder.forInputStreamReader(StreamDecoder.java:71) ~[?:1.8.0_45]
        at java.io.InputStreamReader.<init>(InputStreamReader.java:100) ~[?:1.8.0_45]
        at com.sun.mail.handlers.text_plain.getContent(text_plain.java:80) ~[javax.mail-1.5.6.jar:1.5.6]
        at javax.activation.DataSourceDataContentHandler.getContent(DataHandler.java:795) ~[?:1.8.0_45]
        at javax.activation.DataHandler.getContent(DataHandler.java:542) ~[?:1.8.0_45]
        at javax.mail.internet.MimeBodyPart.getContent(MimeBodyPart.java:657) ~[javax.mail-1.5.6.jar:1.5.6]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:124) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:134) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.runIfMatches(EmailEcaRule.groovy:84) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceFacadeImpl.runEmecaRules(ServiceFacadeImpl.groovy:429) ~[moqui-framework-2.0.0.jar:2.0.0]
        at pollEmailServer_groovy.run(pollEmailServer_groovy:89) ~[script:?]
        at org.moqui.impl.context.runner.GroovyScriptRunner.run(GroovyScriptRunner.groovy:50) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.ResourceFacadeImpl.script(ResourceFacadeImpl.groovy:356) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.runner.ScriptServiceRunner.runService(ScriptServiceRunner.groovy:49) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:317) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
11:33:57.563  WARN oquiWorker-3      o.moqui.i.c.TransactionFacadeImpl Transaction rollback for [Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)]. Here is the current location: 
org.moqui.BaseException: Rollback location
        at org.moqui.impl.context.TransactionFacadeImpl.rollback(TransactionFacadeImpl.groovy:425) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.TransactionFacadeImpl.rollback(TransactionFacadeImpl.groovy:408) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:340) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
11:33:57.625  WARN oquiWorker-3        o.moqui.i.s.ServiceCallSyncImpl Error running service [org.moqui.impl.EmailServices.poll#EmailServer] (Throwable)
java.io.UnsupportedEncodingException: unicode-1-1-utf-7
        at sun.nio.cs.StreamDecoder.forInputStreamReader(StreamDecoder.java:71) ~[?:1.8.0_45]
        at java.io.InputStreamReader.<init>(InputStreamReader.java:100) ~[?:1.8.0_45]
        at com.sun.mail.handlers.text_plain.getContent(text_plain.java:80) ~[javax.mail-1.5.6.jar:1.5.6]
        at javax.activation.DataSourceDataContentHandler.getContent(DataHandler.java:795) ~[?:1.8.0_45]
        at javax.activation.DataHandler.getContent(DataHandler.java:542) ~[?:1.8.0_45]
        at javax.mail.internet.MimeBodyPart.getContent(MimeBodyPart.java:657) ~[javax.mail-1.5.6.jar:1.5.6]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:124) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.makeBodyPartList(EmailEcaRule.groovy:134) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.EmailEcaRule.runIfMatches(EmailEcaRule.groovy:84) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceFacadeImpl.runEmecaRules(ServiceFacadeImpl.groovy:429) ~[moqui-framework-2.0.0.jar:2.0.0]
        at pollEmailServer_groovy.run(pollEmailServer_groovy:89) ~[script:?]
        at org.moqui.impl.context.runner.GroovyScriptRunner.run(GroovyScriptRunner.groovy:50) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.context.ResourceFacadeImpl.script(ResourceFacadeImpl.groovy:356) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.runner.ScriptServiceRunner.runService(ScriptServiceRunner.groovy:49) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.callSingle(ServiceCallSyncImpl.groovy:317) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallSyncImpl.call(ServiceCallSyncImpl.groovy:165) ~[moqui-framework-2.0.0.jar:2.0.0]
        at org.moqui.impl.service.ServiceCallJobImpl$ServiceJobCallable.call(ServiceCallJobImpl.groovy:207) [moqui-framework-2.0.0.jar:2.0.0]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_45]
```

Header information and content from one of the problem messages:

```
Received: from webmail8.myit.local (LHLO webmail8.xxxxxxxxxxxx.ie)
 (10.1.0.19) by webmail8.xxxxxxxxxxxx.ie with LMTP; Mon, 14 Sep 2015
 12:16:27 +0100 (IST)
Received: from localhost (localhost [127.0.0.1])
    by webmail8.xxxxxxxxxxxx.ie (Postfix) with ESMTP id 417563D1DF2
    for <xxxxxx.xxxxxx@xxxxxxx.xx>; Mon, 14 Sep 2015 12:16:27 +0100 (IST)
X-Spam-Flag: NO
X-Spam-Score: -3.406
X-Spam-Level:
X-Spam-Status: No, score=-3.406 tagged_above=-10 required=6.6
    tests=[BAYES_00=-1.9, HTML_MESSAGE=0.001, RCVD_IN_DNSWL_MED=-2.3,
    RDNS_NONE=0.793] autolearn=ham autolearn_force=no
Received: from webmail8.xxxxxxxxxxxx.ie ([127.0.0.1])
    by localhost (webmail8.xxxxxxxxxxxx.ie [127.0.0.1]) (amavisd-new, port 10032)
    with ESMTP id qD6OPal5dRxW for <xxxxx.xxxxx@xxxxxxx.xx>;
    Mon, 14 Sep 2015 12:16:24 +0100 (IST)
Received: from localhost (localhost [127.0.0.1])
    by webmail8.xxxxxxxxxxxx.ie (Postfix) with ESMTP id 55D38EB089
    for <xxxxx.xxxxx@xxxxxxx.xx>; Mon, 14 Sep 2015 12:16:24 +0100 (IST)
X-Virus-Scanned: amavisd-new at webmail8.xxxxxxxxxxxx.ie
Received: from webmail8.xxxxxxxxxxxx.ie ([127.0.0.1])
    by localhost (webmail8.xxxxxxxxxxxx.ie [127.0.0.1]) (amavisd-new, port 10026)
    with ESMTP id L5y00C_28rjW for <xxxxx.xxxxx@xxxxxxx.xx>;
    Mon, 14 Sep 2015 12:16:23 +0100 (IST)
Received: from incoming2.xxxxxxxxxxxx.ie (myitgfi2.myit.local [10.1.0.7])
    by webmail8.xxxxxxxxxxxx.ie (Postfix) with ESMTP id CEB223E63D6
    for <xxxxx.xxxxx@xxxxxxx.xx>; Mon, 14 Sep 2015 12:16:23 +0100 (IST)
Received: from mail.xxxxx.xx ([213.190.140.194]) by incoming2.xxxxxxxxxxxx.ie with Microsoft SMTPSVC(6.0.3790.4675);
     Mon, 14 Sep 2015 12:16:11 +0100
Received: from mail.xxxxx.xx (localhost.localdomain [127.0.0.1])
    by localhost (Email Security Appliance) with SMTP id C174BA7AC1_5F6AC7BB
    for <xxxxx.xxxxx@xxxxxxx.xx>; Mon, 14 Sep 2015 11:16:11 +0000 (GMT)
Received: from UDG-EXCCAS01.ud.plc (unknown [172.17.27.103])
    (using TLSv1 with cipher AES128-SHA (128/128 bits))
    (Client CN "UDG-EXCCAS01", Issuer "UDG-EXCCAS01" (not verified))
    by mail.xxxxx.xx (Sophos Email Appliance) with ESMTPS id 94B58A7ABD_5F6AC7BF
    for <xxxxx.xxxxx@xxxxxxx.xx>; Mon, 14 Sep 2015 11:16:11 +0000 (GMT)
MIME-Version: 1.0
From: <postmaster@xxxxx.xx>
To: <xxxxx.xxxxx@xxxxxxx.xx>
Date: Mon, 14 Sep 2015 12:16:11 +0100
Content-Type: multipart/report; report-type=delivery-status;
    boundary="163d2f30-7ff5-4e62-9b98-b857cd0b3e23"
Content-Language: en-IE
Message-ID: <dd2d8e21-8649-491b-bdf5-419ba32a62ca@xxxxx.xx>
In-Reply-To: <X2636D44GT1UBALJ1U5ON@xxxxxxx.xx>
References: <X2636D44GT1UBALJ1U5ON@xxxxxxx.xx>
Subject: Undeliverable: Attached sales note 
X-OriginalArrivalTime: 14 Sep 2015 11:16:11.0991 (UTC) FILETIME=[C3440670:01D0EEDE]


--163d2f30-7ff5-4e62-9b98-b857cd0b3e23
Content-Type: multipart/alternative; differences=Content-Type;
    boundary="9479a5bd-66fa-4aa2-a2c2-fcef081f5754"


--9479a5bd-66fa-4aa2-a2c2-fcef081f5754
Content-Type: text/plain; charset="us-ascii"
Content-Transfer-Encoding: quoted-printable

Delivery has failed to these recipients or groups:


Diagnostic information for administrators:

Generating server: xxxxx.xx

xxxxxxxx.xxxxxx@xxxxx.xxx
#550 5.1.1 RESOLVER.ADR.RecipNotFound; not found ##rfc822;xxxxxxxx.xxxxxx=
@xxxxx.xxx

Original message headers:

Received: from mail.xxxxx.xx (172.17.91.12) by UDG-EXCCAS01.UD.PLC
 (172.17.27.110) with Microsoft SMTP Server (TLS) id 14.3.123.3; Mon, 14 Sep
 2015 12:16:11 +0100
Received: from mail.xxxxx.xx (localhost.localdomain [127.0.0.1])  by
 localhost (Email Security Appliance) with SMTP id 32134A7AB2_5F6AC7BB  for
 <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep 2015 11:16:11 +0000 (GMT)
Received: from incoming2.xxxxxxxxxxxx.ie
 (host225-244-static.99-5-b.business.telecomitalia.it [5.99.244.225])   by
 mail.xxxxx.xx (Sophos Email Appliance) with ESMTP id
 CCD7BA7AAA_5F6AC78F    for <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep =
2015
 11:16:08 +0000 (GMT)
Received: by mail.united-drug.com with SMTP id S7170        for
 <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep 2015 11:11:07 +0058 CEST
MIME-Version: 1.0
X-Received: by 213.190.140.194 with SMTP id 388QRO7VTKI4V6CE;
 Mon, 14 Sep 2015 11:11:07 +0058 CEST
Received: by 213.190.140.194 with HTTP; Mon, 14 Sep 2015 11:11:07 +0058 CEST
Date: Mon, 14 Sep 2015 11:11:07 +0058
Message-ID: <X2636D44GT1UBALJ1U5ON@xxxxxxx.xx>
Subject: Attached sales note
From: xxxxx <xxxxx.xxxxx@xxxxxxx.xx>
To: <xxxxxxxx.xxxxxx@xxxxx.xxx>
Content-Type: multipart/mixed;
        boundary=3D"2PW33G54863183M46JUX432TR072WH012HV8"
X-Sophos-SenderHistory: ip=3D5.99.244.225,fs=3D0,da=3D11880510,mc=3D1
X-Sophos-ESA: [mail.xxxxx.xx] 3.9.0.0, Antispam-Engine: 2.7.2.1390750=
, Antispam-Data: 2015.9.14.110617
X-SEA-Spam: Gauge=3DXXXIII, Probability=3D33%, Report=3D'
 ZIP_ATTACHED_EXE 3.5, HTML_00_01 0.05, HTML_00_10 0.05, BODYTEXTP_SIZE_300=
0_LESS 0, BODY_SIZE_10000_PLUS 0, DQ_SUSP_1 0, DQ_SUSP_2 0, DQ_SUSP_4 0, NO=
_URI_FOUND 0, NO_URI_HTTPS 0, RDNS_STATIC 0, SPF_FAIL 0, WEBMAIL_SOURCE 0, =
ZIP_ATTACHED 0, __CT 0, __CTYPE_HAS_BOUNDARY 0, __CTYPE_MULTIPART 0, __CTYP=
E_MULTIPART_MIXED 0, __DQ_HEUR_4 0, __DQ_HEUR_5 0, __DQ_IP_SUSP_1 0, __DQ_I=
P_SUSP_2 0, __DQ_S_HIST_1 0, __DQ_S_IP_100K 0, __DQ_S_IP_10K 0, __DQ_S_IP_1=
K 0, __HAS_FROM 0, __HAS_MSGID 0, __MIME_VERSION 0, __PHISH_SPEAR_GREETING =
0, __PHISH_SPEAR_HTTP_RECEIVED 0, __RDNS_STATIC_1 0, __SANE_MSGID 0, __SUBJ=
_ALPHA_END 0, __TO_MALFORMED_2 0, __TO_NO_NAME 0, __ZIP_ATTACHED 0, __ZIP_A=
TTACHED2 0, __ZIP_ATTACHED_EXE 0, __ZIP_ATTACHED_EXE2 0'
Return-Path: xxxxx.xxxxx@xxxxxxx.xx


The information in this email is confidential and may be legally privileged=
. It is intended solely for the addressee. Access to this email by anyone e=
lse is unauthorised. If you are not the intended recipient, any disclosure,=
 copying, distribution or any action taken or omitted to be taken in relian=
ce on it, is prohibited and may be unlawful. All views, opinions or advice =
contained in this communication are those of the sender and not necessarily=
 those of xxxxxXXXXXXX plc or its affiliate. xxxxxXXXXXXX plc or its af=
filiate cannot guarantee that this e-mail and attachments are free from vir=
uses and you must ensure that you carry out your own virus checks. UDG Heal=
thcare plc or its affiliate accepts no liability for any loss or damage cau=
sed by software viruses.

ADDRESS REMOVED



--9479a5bd-66fa-4aa2-a2c2-fcef081f5754
Content-Type: text/html; charset="us-ascii"
Content-Transfer-Encoding: quoted-printable

<html>
<Head></head><body>
<p><b><font color=3D"#000066" size=3D"3" face=3D"Arial">Delivery has failed=
 to these recipients or groups:</font></b></p>
<font color=3D"#000000" size=3D"2" face=3D"Tahoma"><p><a href=3D"mailto:sio=
bhan.geoghegan@xxxxx.xxx">xxxxxxxx.xxxxxx@xxxxx.xxx</a><br>
The e-mail address you entered couldn't be found. Please check the recipien=
t's e-mail address and try to resend the message. If the problem continues,=
 please contact your helpdesk.<br>
</p>
</font>
<br><br><br><br><br><br>
<font color=3D"#808080" size=3D"2" face=3D"Tahoma"><p><b>Diagnostic informa=
tion for administrators:</b></p>
<p>Generating server: xxxxx.xx</p>
<p>xxxxxxxx.xxxxxx@xxxxx.xxx<br>
 #550 5.1.1 RESOLVER.ADR.RecipNotFound; not found ##rfc822;siobhan.geoghega=
n@xxxxx.xxx</p>
<p>Original message headers:</p>
<pre>Received: from mail.xxxxx.xx (172.17.91.12) by UDG-EXCCAS01.UD.P=
LC
 (172.17.27.110) with Microsoft SMTP Server (TLS) id 14.3.123.3; Mon, 14 Sep
 2015 12:16:11 +0100
Received: from mail.xxxxx.xx (localhost.localdomain [127.0.0.1])    by
 localhost (Email Security Appliance) with SMTP id 32134A7AB2_5F6AC7BB  for
 &lt;xxxxxxxx.xxxxxx@xxxxx.xxx&gt;; Mon, 14 Sep 2015 11:16:11 +0000 (=
GMT)
Received: from incoming2.xxxxxxxxxxxx.ie
 (host225-244-static.99-5-b.business.telecomitalia.it [5.99.244.225])   by
 mail.xxxxx.xx (Sophos Email Appliance) with ESMTP id
 CCD7BA7AAA_5F6AC78F    for &lt;xxxxxxxx.xxxxxx@xxxxx.xxx&gt;; Mon, 14 S=
ep 2015
 11:16:08 +0000 (GMT)
Received: by mail.united-drug.com with SMTP id S7170        for
 &lt;xxxxxxxx.xxxxxx@xxxxx.xxx&gt;; Mon, 14 Sep 2015 11:11:07 +0058 C=
EST=20
MIME-Version: 1.0
X-Received: by 213.190.140.194 with SMTP id 388QRO7VTKI4V6CE;
 Mon, 14 Sep 2015 11:11:07 +0058 CEST=20
Received: by 213.190.140.194 with HTTP; Mon, 14 Sep 2015 11:11:07 +0058 CES=
T=20
Date: Mon, 14 Sep 2015 11:11:07 +0058
Message-ID: &lt;X2636D44GT1UBALJ1U5ON@xxxxxxx.xx&gt;
Subject: Attached sales note=20
From: xxxxx &lt;xxxxx.xxxxx@xxxxxxx.xx&gt;
To: &lt;xxxxxxxx.xxxxxx@xxxxx.xxx&gt;
Content-Type: multipart/mixed;
    boundary=3D&quot;2PW33G54863183M46JUX432TR072WH012HV8&quot;
X-Sophos-SenderHistory: ip=3D5.99.244.225,fs=3D0,da=3D11880510,mc=3D1
X-Sophos-ESA: [mail.xxxxx.xx] 3.9.0.0, Antispam-Engine: 2.7.2.1390750=
, Antispam-Data: 2015.9.14.110617
X-SEA-Spam: Gauge=3DXXXIII, Probability=3D33%, Report=3D'
 ZIP_ATTACHED_EXE 3.5, HTML_00_01 0.05, HTML_00_10 0.05, BODYTEXTP_SIZE_300=
0_LESS 0, BODY_SIZE_10000_PLUS 0, DQ_SUSP_1 0, DQ_SUSP_2 0, DQ_SUSP_4 0, NO=
_URI_FOUND 0, NO_URI_HTTPS 0, RDNS_STATIC 0, SPF_FAIL 0, WEBMAIL_SOURCE 0, =
ZIP_ATTACHED 0, __CT 0, __CTYPE_HAS_BOUNDARY 0, __CTYPE_MULTIPART 0, __CTYP=
E_MULTIPART_MIXED 0, __DQ_HEUR_4 0, __DQ_HEUR_5 0, __DQ_IP_SUSP_1 0, __DQ_I=
P_SUSP_2 0, __DQ_S_HIST_1 0, __DQ_S_IP_100K 0, __DQ_S_IP_10K 0, __DQ_S_IP_1=
K 0, __HAS_FROM 0, __HAS_MSGID 0, __MIME_VERSION 0, __PHISH_SPEAR_GREETING =
0, __PHISH_SPEAR_HTTP_RECEIVED 0, __RDNS_STATIC_1 0, __SANE_MSGID 0, __SUBJ=
_ALPHA_END 0, __TO_MALFORMED_2 0, __TO_NO_NAME 0, __ZIP_ATTACHED 0, __ZIP_A=
TTACHED2 0, __ZIP_ATTACHED_EXE 0, __ZIP_ATTACHED_EXE2 0'
Return-Path: xxxxx.xxxxx@xxxxxxx.xx
</pre>
</font>
<PRE>The information in this email is confidential and may be legally privi=
leged. It is intended solely for the addressee. Access to this email by any=
one else is unauthorised. If you are not the intended recipient, any disclo=
sure, copying, distribution or any action taken or omitted to be taken in r=
eliance on it, is prohibited and may be unlawful. All views, opinions or ad=
vice contained in this communication are those of the sender and not necess=
arily those of xxxxxXXXXXXX plc or its affiliate. xxxxxXXXXXXX plc or i=
ts affiliate cannot guarantee that this e-mail and attachments are free fro=
m viruses and you must ensure that you carry out your own virus checks. UDG=
 Healthcare plc or its affiliate accepts no liability for any loss or damag=
e caused by software viruses.

ADDRESS INFO REMOVED

</PRE></body>
</html>=

--9479a5bd-66fa-4aa2-a2c2-fcef081f5754--

--163d2f30-7ff5-4e62-9b98-b857cd0b3e23
Content-Type: message/delivery-status

Reporting-MTA: dns;xxxxx.xx
Received-From-MTA: dns;mail.xxxxx.xx
Arrival-Date: Mon, 14 Sep 2015 11:16:11 +0000

Original-Recipient: rfc822;xxxxxxxx.xxxxxx@xxxxx.xxx
Final-Recipient: rfc822;xxxxxxxx.xxxxxx@xxxxx.xxx
Action: failed
Status: 5.1.1
Diagnostic-Code: smtp;550 5.1.1 RESOLVER.ADR.RecipNotFound; not found


--163d2f30-7ff5-4e62-9b98-b857cd0b3e23
Content-Type: message/rfc822

Received: from mail.xxxxx.xx (172.17.91.12) by UDG-EXCCAS01.UD.PLC
 (172.17.27.110) with Microsoft SMTP Server (TLS) id 14.3.123.3; Mon, 14 Sep
 2015 12:16:11 +0100
Received: from mail.xxxxx.xx (localhost.localdomain [127.0.0.1])    by
 localhost (Email Security Appliance) with SMTP id 32134A7AB2_5F6AC7BB  for
 <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep 2015 11:16:11 +0000 (GMT)
Received: from incoming2.xxxxxxxxxxxx.ie
 (host225-244-static.99-5-b.business.telecomitalia.it [5.99.244.225])   by
 mail.xxxxx.xx (Sophos Email Appliance) with ESMTP id
 CCD7BA7AAA_5F6AC78F    for <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep 2015
 11:16:08 +0000 (GMT)
Received: by mail.united-drug.com with SMTP id S7170        for
 <xxxxxxxx.xxxxxx@xxxxx.xxx>; Mon, 14 Sep 2015 11:11:07 +0058 CEST 
MIME-Version: 1.0
X-Received: by 213.190.140.194 with SMTP id 388QRO7VTKI4V6CE;
 Mon, 14 Sep 2015 11:11:07 +0058 CEST 
Received: by 213.190.140.194 with HTTP; Mon, 14 Sep 2015 11:11:07 +0058 CEST 
Date: Mon, 14 Sep 2015 11:11:07 +0058
Message-ID: <X2636D44GT1UBALJ1U5ON@xxxxxxx.xx>
Subject: Attached sales note 
From: xxxxx <xxxxx.xxxxx@xxxxxxx.xx>
To: <xxxxxxxx.xxxxxx@xxxxx.xxx>
Content-Type: multipart/mixed;
    boundary="2PW33G54863183M46JUX432TR072WH012HV8"
X-Sophos-SenderHistory: ip=5.99.244.225,fs=0,da=11880510,mc=1
X-Sophos-ESA: [mail.xxxxx.xx] 3.9.0.0, Antispam-Engine: 2.7.2.1390750, Antispam-Data: 2015.9.14.110617
X-SEA-Spam: Gauge=XXXIII, Probability=33%, Report='
 ZIP_ATTACHED_EXE 3.5, HTML_00_01 0.05, HTML_00_10 0.05, BODYTEXTP_SIZE_3000_LESS 0, BODY_SIZE_10000_PLUS 0, DQ_SUSP_1 0, DQ_SUSP_2 0, DQ_SUSP_4 0, NO_URI_FOUND 0, NO_URI_HTTPS 0, RDNS_STATIC 0, SPF_FAIL 0, WEBMAIL_SOURCE 0, ZIP_ATTACHED 0, __CT 0, __CTYPE_HAS_BOUNDARY 0, __CTYPE_MULTIPART 0, __CTYPE_MULTIPART_MIXED 0, __DQ_HEUR_4 0, __DQ_HEUR_5 0, __DQ_IP_SUSP_1 0, __DQ_IP_SUSP_2 0, __DQ_S_HIST_1 0, __DQ_S_IP_100K 0, __DQ_S_IP_10K 0, __DQ_S_IP_1K 0, __HAS_FROM 0, __HAS_MSGID 0, __MIME_VERSION 0, __PHISH_SPEAR_GREETING 0, __PHISH_SPEAR_HTTP_RECEIVED 0, __RDNS_STATIC_1 0, __SANE_MSGID 0, __SUBJ_ALPHA_END 0, __TO_MALFORMED_2 0, __TO_NO_NAME 0, __ZIP_ATTACHED 0, __ZIP_ATTACHED2 0, __ZIP_ATTACHED_EXE 0, __ZIP_ATTACHED_EXE2 0'
Return-Path: xxxxx.xxxxx@xxxxxxx.xx


--2PW33G54863183M46JUX432TR072WH012HV8
Content-Type: text/plain; charset="utf-8"; format=flowed
Content-Transfer-Encoding: 7bit

Potentially dangerous attachments were identified and removed from this
message.

From: "xxxxx" <xxxxx.xxxxx@xxxxxxx.xx>
To: xxxxxxxx.xxxxxx@xxxxx.xxx
Subject: Attached sales note 
When: Mon, 14 Sep 2015 11:11:07 +0058 CEST

If you feel this is an error and wish this email to be released, please reply to this email leaving the message intact. The message (excluding the attachment) is included below.



Dear customer,

Please find your sales order #57650  attached.

xxxxx

--2PW33G54863183M46JUX432TR072WH012HV8--

--163d2f30-7ff5-4e62-9b98-b857cd0b3e23--
```

Additional Information: This is a 'Zimbra' type mail server being polled. One note I made in researching this problem was...  _"The IMAP BODYSTRUCTURE response includes the exact size of each body part. Normally, this size is used to determine how much data to fetch for each body part. **Some servers report this size incorrectly in some cases; this property can be set to work around such server bugs.** If this property is set to true, this size is ignored and data is fetched until the server reports the end of data. This will result in an extra fetch if the data size is a multiple of the block size. Defaults to false."_ (From https://javamail.java.net/nonav/docs/api/com/sun/mail/imap/package-summary.html).

**Comment by jonesde** [comment]
About your last comment, have you tried setting the mail.imap.ignorebodystructuresize property to see if that fixes it?

Note that the pollEmailServer.groovy script passes in System properties for the mail session properties, so you can set a system property for this to avoid changing the code. As a convenience this can be done in your runtime or component Moqui Conf XML file using the default-property element (goes near the top, see MoquiDefaultConf.xml for examples).

BTW, this looks more like a specific issue with Zimbra and the javax.mail library than with Moqui... it may be useful for other Moqui users but is sort of off-topic (ie chances are there is nothing we can do in Moqui to fix this sort of issue, an issue with libraries used in Moqui).

**Comment by ronankeane** [comment]
No, that didn't fix it. (I put the following in the conf file:)

`<default-property name="mail.imap.ignorebodystructuresize" value="true"/>`

I understand what you are saying about it being a specific issue with the email server and the javax.mail library. 

On this (for reference), I found some advice from the author of the library (that I think applies to this case)  saying as a 'workaround' to use the MimeMessage copy constructor to make a copy of the entire message - fetching the entire message from the server and parsing it on the client to access the email content, and then use the original message object to setFlags and so on, on the mail server.

It appears it may be a common problem with a variety of email servers that don't bother to fully implement the IMAP protocol.

**Comment by jonesde** [comment]
I think I understand what you mean by copying the MimeMessage object, made that change in commit #e69de9c. I also changed it to log the properties in the mail session so you can see if your property setting made it all the way through.

It would be great to find a solution to this, chances are others will run into it.

**Comment by ronankeane** [comment]
I'm having trouble testing this solution at the moment. I will look into other ways to do it.

The reason for this is that my use case is to "harvest" emails as an adjunct to the regular email client (Outlook) to track quantity of communications over periods with "subscriber" organisations, and track communications per contact/organisation/initiative etc. Some mailboxes go back years, with large attachments etc. 

The mailbox I'm currently testing, with these problem mails, is one such. **Thus, in this case it is not practical to download the whole message _before_ the filtering through EMECA rules, and so on.**

I understand from the error log that the error is being generated at line 124 of EmailEcaRule.groovy.
`Object content = part.getContent()` [http://docs.oracle.com/javaee/6/api/javax/mail/internet/MimeMessage.html#getContent()]

So I thought for my own use case, since the error is very rare (though unrecoverable at present), that I would simply return a generic content object with a simple string explaining the error, on a  java.io.UnsupportedEncodingException or  java.io.UnsupportedEncodingException. (It also occurred to me that perhaps a better solution _at_ this error would be to go back and then use copy constructor to download the message and try again, but I am not sure if this is feasible or makes sense).

This is just my perspective as a novice programmer. I have not tried to implement and test the above as yet. 

Certainly, the solution proposed above in the new commit makes most sense for most regular uses cases though. I will endeavour to test it soon and revert.

**Comment by ronankeane** [comment]
As an aside, for reference (hopefully others might find it useful) - since in my own use case I am uninterested in storing attachments (only need to extract the message content in simple text format), the intention is to use the following section of code in my service to extract from BodyPartList. - Inspired and derived from this code [here](https://gusto77.wordpress.com/2012/04/03/extracting-text-from-email-message-with-javamail/)).  

**Note the code below was first written to use directly in EmailECARule.groovy, but on consideration obviously, I should keep editing of framework files to an absolute minimum.** 

So _the intention is to re-write again to iterate through bodyPartList in a similar manner _in my xml service._ 

But included here anyway for reference. (uses [JSoup html parser](https://jsoup.org/) - currently html markup is output for some contentTypes in the existing Moqui services).

```
else if (content instanceof Multipart) {
                BodyPart clearTextPart = null
                BodyPart htmlTextPart = null
                Multipart mpContent = (Multipart) content
                int count = mpContent.getCount()
                for (int i = 0; i < count; i++) {
                    BodyPart bp =  mpContent.getBodyPart(i)
                    if(bp.isMimeType("text/plain")) {
                        clearTextPart = bp
                        break;
                    }
                    else if(bp.isMimeType("text/html")) {
                        htmlTextPart = bp
                    }
                }
                if(clearTextPart != null) {
                    bpMap = (String) clearTextPart.getContent()
                }
                else if (htmlTextPart!=null) {
                    String html = (String) htmlTextPart.getContent()
                    bpMap = Jsoup.parse(html).text()
                }
				bodyPartList.add(bpMap)
        } 
```

**Comment by ronankeane** [comment]
This was eventually resolved by setting a number of the **system properties** from here http://docs.oracle.com/javaee/6/api/javax/mail/internet/package-summary.html to deal with various problems in emails.

I'm sorry I was not able to test the above fix, as the problem emails were dated 2 years ago in a live employee account and I would have had to download everything in the mailbox to get to it. (With the above modification, it is much much slower to sift through thousands of emails). 

For reference, initially, I tried using the EMECA rules to skip the problem emails with the use of regex. 

`<condition><expression>fields.from =~ /^((?!(M|m)ailer).)*$/ &amp;&amp; fields.from =~ /^((?!(P|p)ostmaster).)*$/ &amp;&amp; fields.subject =~ /^((?!(U|u)ndeliverable).)*$/</expression></condition>`

That didn't work. - Although I read the error log to mean that the 'makeBodyPartList' part of EmailEcaRule.groovy is only run after the condition passes, the execution did not seem to reach the logger.info("======== EMECA... line in EmailEcaRule.groovy. I.e. It seemed to fail before that logger information. I don't quite understand this.

Just to note too - I tried to use a date condition on fields.sentDate in the Emeca. However, I was getting an error, 'java.lang.String cannot be cast to java.util.Date'. (Instead I put an equivalent 'filter' directly in my xml service, a tailored equivalent of the standard save#EcaEmailMessage service in Moqui.).

**Comment by ronankeane** [comment]
Each 'run' taking a long time to go through emails before I got to the problem ones, caused me to get frustrated and I'm afraid I just threw a number of property settings at the problem en masse. They were:
mail.mime.base64.ignoreerrors
mail.mime. ignoreunknownencoding
mail.mime.uudecode. ignoreerrors
mail.mime.multipart. allowempty
mail.mime.multipart. ignoreexistingboundaryparameter
Wrt to the condition in the EMECA - my problem was the need to calculate the 'cutOffDate' in a script before the condition. I assume you cannot do this before the condition tag in an EMECA.

**Comment by jonesde** [comment]
How many emails are we talking about?

I can imagine this might not go well... this code was really designed for getting new email incrementally as opposed to going through archives. Admittedly it is not the most efficient as it tries to grab all the email data and put it in structures that are little easier to use for common cases, so everything is available for the EMECA rule conditions, etc. As such it isn't all that efficient for filtering through a bunch of email.

This could maybe be improved, but it sounds like what you really need is something different and perhaps based on the pollEmailServer code but different in that you process the MimeMessage objects directly instead of running them through the EMECA stuff.

On the framework level some sort of lower level service that does call backs might be useful, like something to just use the EmailServer entity to get the mail Store, Folder, etc and just return those without doing anything else (not even opening the folder or searching it)... then you can do more specific searches on the folder and so on. If that is of interest please create a new issue (feature request sort of thing).

Anyway, is there anything else you see that can be done or needs to be done related to this issue or can it be closed?

**Comment by ronankeane** [comment]
Thanks, yes, this issue was fully resolved. Thank you for the support.

PS The overall functionality was achieved, albeit, there are constraints on numbers of emails that can be processsed due mainly to speed of access. (By deploying an active email management and archiving strategy in parallel within this SME organisation, this is ok.)... Also, it was required to use more heap space (and physical memory), and for the moment I am passing UseParallelOldGC and -UseGCOverheadLimit arguments to the JVM'. As the polling happens only at night, application pauses are acceptable. But I am continuing to experiment with garbage collection and heap settings. 

Yes, I will put together a feature request based on my experience, thank you for the suggestion. I guess the use of 'SearchTerm' is what is required - However I try to will firm up on this and revert with a new issue soon.
